pkg/tun/tun.go

```diff
@@ -0,0 +1,164 @@
+	return func(p PacketHeader) bool {
+		return p.DestinationAddr != ip
+	}
+}
+
+// Device wraps wg tunnel device.
+type Device struct {
+	Tun           tun.Device
+	PacketFilters []PacketFilter
+}
+
+// CreateTUN creates a Device with the provided name and MTU.
+func CreateTUN(iface string, mtu int, packetFilters ...PacketFilter) (*Device, error) {
+	tun, err := tun.CreateTUN(iface, mtu)
+	if err != nil {
+		return nil, fmt.Errorf("error creating tun device: %w", err)
+	}
+
+	return &Device{Tun: tun, PacketFilters: packetFilters}, nil
+}
+
+// File returns the file descriptor of the device.
+func (d *Device) File() *os.File {
+	return d.Tun.File()
+}
+
+// Read one or more packets from the Device (without any additional headers).
+// On a successful read it returns the number of packets read, and sets
+// packet lengths within the sizes slice. len(sizes) must be >= len(bufs).
+// A nonzero offset can be used to instruct the Device on where to begin
+// reading into each element of the bufs slice.
+func (d *Device) Read(bufs [][]byte, sizes []int, offset int) (int, error) {
+	return d.Tun.Read(bufs, sizes, offset)
+}
+
+// Write one or more packets to the device (without any additional headers).
+// On a successful write it returns the number of packets written. A nonzero
+// offset can be used to instruct the Device on where to begin writing from
+// each packet contained within the bufs slice.
+func (d *Device) Write(bufs [][]byte, offset int) (int, error) {
+	if len(d.PacketFilters) > 0 {
+		result := bufs[:0]
+
+		for _, buf := range bufs {
+			packet, err := d.decodePacketHeader(buf)
+			if err != nil {
+				return 0, err
+			}
+
+			var drop bool
+
+			for _, filter := range d.PacketFilters {
+				if filter(packet) {
+					drop = true
+
+					break
+				}
+			}
+
+			if !drop {
+				result = append(result, buf)
+			}
+
+			bufs = result
+		}
+	}
+
+	return d.Tun.Write(bufs, offset)
+}
+
+// MTU returns the MTU of the Device.
+func (d *Device) MTU() (int, error) {
+	return d.Tun.MTU()
+}
+
+// Name returns the current name of the Device.
+func (d *Device) Name() (string, error) {
+	return d.Tun.Name()
+}
+
+// Events returns a channel of type Event, which is fed Device events.
+func (d *Device) Events() <-chan tun.Event {
+	return d.Tun.Events()
+}
+
+// Close stops the Device and closes the Event channel.
+func (d *Device) Close() error {
+	return d.Tun.Close()
+}
+
+// BatchSize returns the preferred/max number of packets that can be read or
+// written in a single read/write call. BatchSize must not change over the
+// lifetime of a Device.
+func (d *Device) BatchSize() int {
+	return d.Tun.BatchSize()
+}
+
+func (*Device) decodePacketHeader(data []byte) (PacketHeader, error) {
+	if len(data) < 54 {
+		return PacketHeader{}, fmt.Errorf("packet too short to be a valid IPv6 header")
+	}
+
+	data = data[16:]
+
+	packetHeader := PacketHeader{}
+	packetHeader.Version = data[0] >> 4
+
+	if packetHeader.Version != 6 {
+		return packetHeader, nil
```

Comment:
it should be invalid if not IPv6? or we should also state this explicitly

Reply:
I thought that for invalid ipv6 we can just drop them